usb: device_next: add BT HCI USB transport layer

jfischer-no opened the pull request:
Add Bluetooth HCI USB transport layer implementation for the new USB device support,
porting of subsys/usb/device/class/bluetooth.c without H4 bloat.

jfischer-no committed:
usb: device_next: add BT HCI USB transport layer

Add Bluetooth HCI USB transport layer implementation for the new
experimental USB device support based on the existing implementation
subsys/usb/device/class/bluetooth.c.

This implementation, unlike existing one, contains the interface
descriptors for isochronous endpoints. Since we do not support voice
channels, these are just there to avoid issues with Linux kernel btusb
driver when this implementation is part of a composite configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: hci_usb: enable new experimental USB device support

Enable Bluetooth HCI USB tranport layer from new experimental
USB device support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jhedberg commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+										\
+	USBD_DEFINE_CLASS(bt_hci_##n, &bt_hci_api, &bt_hci_class_##n);
+
+/*
+ * Bluetooth subsystem does not support multiple HCI instances,
+ * but we are almost ready for it.
+ */
+BT_HCI_DESCRIPTOR_DEFINE(0)
+BT_HCI_CLASS_DATA_DEFINE(0)
```
I think this relates to #51503 too. Along with making the HCI driver DTS-based, it becomes more natural to declare multiple instances. That said, the host stack treats the controller as singleton. There's a single `bt_dev` struct that's accessed as a global variable in several places of the host. To make the host truly multi-HCI aware there would need to be this `bt_dev` context per controller instance that's then passed around as a parameter in most existing Bluetooth APIs.

Thalley replied:
@jhedberg @jfischer-no does anything need to be done here?

jfischer-no replied:
As I understood @carlescufi, not in the near future.

jhedberg reviewed (dismissed):
Looks pretty good

Thalley commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -7,12 +7,80 @@
+
+	return 0;
+}
+#endif /* IS_ENABLED(CONFIG_USB_DEVICE_STACK_NEXT) */
```
```suggestion
#endif /* CONFIG_USB_DEVICE_STACK_NEXT */
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
```
Did Intel contribute to this file?

jfischer-no replied:
This is porting of subsys/usb/device/class/bluetooth.c to new experimental USB device support. My fault, I should mention it in commit message.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+ * RAM usage efficient.
+ */
+NET_BUF_POOL_FIXED_DEFINE(bt_hci_ep_pool,
+			  3, 512,
```
Why these specific numbers?

jfischer-no replied:
Three endpoints, maximum three transfers used synchronously, 512 is the maximum packet size of a high speed bulk endpoint. I will add a comment.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+#define USB_BT_HCI_SUBCLASS		0x01
+#define USB_BT_HCI_PROTOCOL		0x01
+
+#define BT_HCI_DEFAULT_BULK_EP_MPS	0	/* Get maximum supported */
```
I don't understand the comment - Can you elaborate?

desowin replied:
The Max Packet Size for bulk endpoints is speed-dependent and device only knows the operating speed after it is connected to host. There is no problem for Full-Speed devices, which always operate as Full-Speed and thus have 64 bytes Max Packet Size for bulk endpoints. However High-Speed devices can operate as both Full-Speed or High-Speed devices depending on connected host. When the stack is parsing endpoint descriptor and sees 0 as max packet size for bulk endpoint, it assigns either 64 (Full-Speed) or 512 (High-Speed).

Thalley replied:
Ah, so when the MPS = `BT_HCI_DEFAULT_BULK_EP_MPS`, then it indicates that we should assigned the maximum supported value?

desowin replied:
"assigned" by the next USB device stack, in the sense that the device stack replaces the value with the max possible at connected speed.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+		struct net_buf *bt_buf;
+		uint8_t ep;
+
+		bt_buf = net_buf_get(&bt_hci_tx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(bt_buf)) {
+		case BT_BUF_EVT:
+			ep = bt_hci_get_int_in(c_nd);
+			break;
+		case BT_BUF_ACL_IN:
+			ep = bt_hci_get_bulk_in(c_nd);
+			break;
+		default:
+			LOG_ERR("Unknown type %u", bt_buf_get_type(bt_buf));
+			continue;
+		}
```
Just to make sure I understand: `in` in this context means from controller to host, right? Or am I misunderstanding what `tx` refers to here?

desowin replied:
In USB context "IN" always refers to device (here: Zephyr device with USB device peripheral) to host (usually PC/laptop, but can be anything host capable).

Thalley replied:
Okay, so `in` is from device to host?

desowin replied:
Yes, `in` means from "USB device" to "USB host". Unfortunately there is also "Bluetooth host" here that can cause confusion when just "host" is used.

Thalley replied:
Yes :) Thanks for the explanation

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+		bt_buf = net_buf_get(&bt_hci_tx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(bt_buf)) {
+		case BT_BUF_EVT:
+			ep = bt_hci_get_int_in(c_nd);
+			break;
+		case BT_BUF_ACL_IN:
+			ep = bt_hci_get_bulk_in(c_nd);
+			break;
```
What about ISO in?

desowin replied:
It is optional and should not be in the first alternate configuration (here: the only one).

Thalley replied:
I'm not sure I understand your comment. ACL is also optional. 

The code even previously mentions ISO, but does not implement anything related to it.

desowin replied:
> I'm not sure I understand your comment. ACL is also optional.

How is it optional? Please direct me to the relevant Bluetooth specification page/chapter.

BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 does not make the ACL Data endpoints optional at all.

After a second read through the Part B: USB Transport Layer it seems that there are two interfaces here:
- First one (the one implemented here) that contains bulk and interrupt endpoints
- Second one that provides scalable isochronous bandwidth

This PR entirely misses the second interface with alternate configurations.

BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B , 2.1.1 Controller descriptors
> The Controller configuration consists of two interfaces. The first interface has
no alternate settings and contains the bulk and interrupt endpoints. The second
interface provides scalable isochronous bandwidth. The recommended
configuration for the second interface has four alternate settings that provide
different bandwidth. The default interface is empty so that the device is capable
of scaling down to zero isochronous bandwidth.

Thalley replied:
> > I'm not sure I understand your comment. ACL is also optional.
> 
> How is it optional? Please direct me to the relevant Bluetooth specification page/chapter.
> 
> BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 does not make the ACL Data endpoints optional at all.

That table (2.1) is a recommendation - Not a specification. If you have a advertising (GAP Broadcaster) or scanning (GAP Observer) device only, you don't have any ACL. 

> After a second read through the Part B: USB Transport Layer it seems that there are two interfaces here:
> 
>     * First one (the one implemented here) that contains bulk and interrupt endpoints
> 
>     * Second one that provides scalable isochronous bandwidth
> 
> 
> This PR entirely misses the second interface with alternate configurations.

Will it be added as part of this PR then, or done later?

desowin replied:
> That table (2.1) is a recommendation - Not a specification. If you have a advertising (GAP Broadcaster) or scanning (GAP Observer) device only, you don't have any ACL.

While there won't be any ACL data in this case, the HCI USB Transport layer (which is implemented here, except for the isochronous interface) must have endpoints for ACL. The text I quoted previously contains
> The first interface has no alternate settings and contains the bulk and interrupt endpoints.

The interrupt endpoint is used for HCI events and bulk endpoints are used for ACL Data (as noted in 2.3 BULK ENDPOINTS EXPECTATIONS). There is no possibility for "interface with only interrupt endpoint".

jfischer-no replied:
> What about ISO in?

What you probably mean will go in the future through bulk endpoint.
This is porting of `subsys/usb/device/class/bluetooth.c ` without h4 bloat (will be added later), nothing more.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+
+		hdr_len = sizeof(*cmd_hdr);
+		cmd_hdr = (struct bt_hci_cmd_hdr *)data;
+		len = cmd_hdr->param_len + hdr_len;
```
Should `len` be `size_t` too? 

We are basically adding a `size_t` and a `uint8_t` and store it in a `uint16_t`; that could cause issues seen from a type perspective. 

Alternatively, should `hdr_len` have been a `uint16_t` or `uint8_t`?

jfischer-no replied:
IIRC, hdr_len must not be smaller uint16_t. But yes it looks weird.

Thalley replied:
Given that it is a struct (`bt_hci_cmd_hdr`) defined by the Core Spec, it won't change, and certainly not a a size that will ever cause issues here. 

If we wanted to be really nitpicking about this, we could add a BUILD_ASSERT to verify, at build time, that this will always work. 

I'm OK with leaving it as is given that the assumptions that this won't cause issue are pretty solid.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+restart_out_transfer:
+	/* TODO: add function to reset transfer buffer and allow to reuse it */
+	net_buf_unref(buf);
+	atomic_clear_bit(&hci_data->state, BT_HCI_ACL_RX_ENGAGED);
+	return bt_hci_acl_out_start(c_nd);
```
```suggestion
	atomic_clear_bit(&hci_data->state, BT_HCI_ACL_RX_ENGAGED);
	
	return bt_hci_acl_out_start(c_nd);
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+		return bt_hci_acl_out_cb(c_nd, buf, err);
+	}
+
+	if (bi->ep == bt_hci_get_bulk_in(c_nd) || bi->ep == bt_hci_get_int_in(c_nd)) {
```
```suggestion
	if (bi->ep == bt_hci_get_bulk_in(c_nd) || 
	    bi->ep == bt_hci_get_int_in(c_nd)) {
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	}
+
+	if (bi->ep == bt_hci_get_bulk_in(c_nd) || bi->ep == bt_hci_get_int_in(c_nd)) {
+		k_sem_give(&hci_data->sync_sem);
+		return 0;
```
```suggestion
		k_sem_give(&hci_data->sync_sem);
		
		return 0;
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+
+	atomic_set_bit(&hci_data->state, BT_HCI_CLASS_ENABLED);
+	LOG_INF("Configuration enabled");
+	if (bt_hci_acl_out_start(c_nd)) {
```
```suggestion
	LOG_INF("Configuration enabled");
	
	if (bt_hci_acl_out_start(c_nd)) {
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+
+	/* We expect host-to-device class request */
+	if (setup->RequestType.type != USB_REQTYPE_TYPE_CLASS) {
+		errno = -ENOTSUP;
+		return 0;
```
```suggestion
		errno = -ENOTSUP;
		
		return 0;
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+
+	cmd_buf = bt_buf_get_tx(BT_BUF_CMD, K_NO_WAIT, buf->data, buf->len);
+	if (!cmd_buf) {
+		LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
```
```suggestion
		LOG_ERR("Cannot get free buffer");
		
		return -ENOMEM;
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	.if0_out_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x02,					\
```
Not familiar with this: Why is this `bEndpointAddress` so low compared to the other 2?

desowin replied:
Because it is OUT, which means from device to host. The addresses 0x80-0x8F are IN, while 0x00-0x0F are OUT.

Universal Serial Bus Specification Revision 2.0, Table 9-13. Standard Endpoint Descriptor
> The address of the endpoint on the USB device described by this descriptor. The address is encoded as follows:
Bit 3...0: The endpoint number
Bit 6...4: Reserved, reset to zero
Bit 7: Direction, ignored for control endpoints
0 = OUT endpoint
1 = IN endpoint

Thalley replied:
I see. 

This could be made clearer with defining the possible endpoint addresses and assigning one of them, instead of these "magic" numbers :)

jfischer-no replied:
There is nothing magic, just basic USB knowledge.

Thalley replied:
No offense, but that is a pretty poor argument :) Firstly the actual value is basically a struct like
```c
struct usb_ep_addr {
    uint8_t number:4,
    uint8_t reserved:3,
    uint8_t direction:1,
};
```

Secondly that argument also implies that the other defined values in this filed also don't need to be defined, as they are presumably also "just basic USB knowledge" :)

jfischer-no replied:
No, until reserved is no longer reserved, what is very unlikely, there is no reason at all to define any macros. 0x81-0x8F | 0x01-0x0F are cross-system and concise in USB context. Also there are USB_CONTROL_EP_IN | USB_CONTROL_EP_OUT macros for 0x80 | 0x00 due to unique and specific nature of these endpoints.

Thalley replied:
> Also there are USB_CONTROL_EP_IN | USB_CONTROL_EP_OUT macros for 0x80 | 0x00 due to unique and specific nature of these endpoints.

Shouldn't these be used then? I'd be happier to see `dir | number` (i.e. `USB_CONTROL_EP_OUT | 0x01`) than just knowing that BIT(7) is implicitly the direction :)

Thalley replied:
Looking at all the other BT USB drivers, they actually define these endpoint addreses such as 
```c
#define BLUETOOTH_INT_EP_ADDR		0x81
#define BLUETOOTH_OUT_EP_ADDR		0x02
#define BLUETOOTH_IN_EP_ADDR		0x82
``` 
or
```c
#define BT_H4_OUT_EP_ADDR               0x01
#define BT_H4_IN_EP_ADDR                0x81
```

and also used in other non-BT drivers such as 
```c

#define RNDIS_INT_EP_ADDR		0x83
#define RNDIS_IN_EP_ADDR		0x82
#define RNDIS_OUT_EP_ADDR		0x01
```

So I still (strongly) request that we use defined values for these

desowin replied:
Multiple USB tools simply show the endpoints in 0x00-0x0F and 0x80-0x8F format. It is really commonplace in USB context to use the hex numbers directly.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+
+#define BT_HCI_DESCRIPTOR_DEFINE(n)						\
+static struct usbd_bt_hci_desc bt_hci_desc_##n = {				\
+	.if0 = {								\
```
Why doesn't this interface have a `bEndpointAddress`?

desowin replied:
Because it is interface descriptor and not an endpoint descriptor. All the endpoints belong to this interface. Notice that there is `bNumEndpoints` here with value 3.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	k_thread_create(&rx_thread_data, rx_thread_stack,
+			K_KERNEL_STACK_SIZEOF(rx_thread_stack),
+			bt_hci_rx_thread, NULL, NULL, NULL,
+			K_PRIO_COOP(8), 0, K_NO_WAIT);
+
+	k_thread_name_set(&rx_thread_data, "bt_hci_rx");
+
+	k_thread_create(&tx_thread_data, tx_thread_stack,
+			K_KERNEL_STACK_SIZEOF(tx_thread_stack),
+			bt_hci_tx_thread, (void *)&bt_hci_0, NULL, NULL,
+			K_PRIO_COOP(8), 0, K_NO_WAIT);
```
Is the priority supposed to be 8 and never anything else? Should it be put into a Kconfig option?

jfischer-no replied:
Yes, I am not sure we need both threads, I have to check wath context bt_send() expects.

Thalley reviewed (changes requested):
Mostly code style comments, and some questions

desowin commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	USBD_VENDOR_REQ(0x00, 0xe0);
+
+struct usbd_bt_hci_desc {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
```
The order is wrong. See BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 Table 2.1 outlines a recommended configuration for a USB Full Speed device

The order should be:
HCI events Interrupt IN
ACL Data Bulk IN
ACL Data Bulk OUT

Moreover, there should also be two dummy isochronous endpoints (first Isoch IN, then Isoch OUT) with Max Packet Size 0 (i.e. no data can be transferred to/from these)

jfischer-no replied:
Order should not matter, as you mention it "outlines a **recommended** configuration", the host have to parse device descriptors and interpret it correctly. I also see no benefit to add isochronous endpoints, there are none in subsys/usb/device/class/bluetooth.c and we do not support classic voice channels.

desowin replied:
As it turns out, Bluetooth specification versions prior to 4.0 had "required configuration" and starting with version 4.0 it is "recommended configuration". For earlier versions the Core Spec Addendum 2 changes the "required" to "recommended".

desowin commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+#define USB_BT_HCI_PROTOCOL		0x01
+
+#define BT_HCI_DEFAULT_BULK_EP_MPS	0	/* Get maximum supported */
+#define BT_HCI_DEFAULT_INT_EP_MPS	64
```
BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 suggests 16 bytes Max Packet Size as the default.

desowin commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,491 @@
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
+} __packed;
```
Atleast add TODO comment about the second interface with 4 alternate settings for isochronous transfers.

jfischer-no committed:
drivers: udc: do not update MPS for isochronous endpoints

Zero data payload size for isochronous endpoints is a
is a valid setting for default interface.
Also do not update MPS of control endpoint since it is
set by the driver.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,507 @@
+		len = sys_le16_to_cpu(acl_hdr->len) + hdr_len;
+		break;
+	}
+	case BT_BUF_ISO_OUT: {
+		struct bt_hci_iso_hdr *iso_hdr;
+
+		hdr_len = sizeof(*iso_hdr);
+		iso_hdr = (struct bt_hci_iso_hdr *)data;
+		len = bt_iso_hdr_len(sys_le16_to_cpu(iso_hdr->len)) + hdr_len;
+		break;
+	}
```
Given that this driver otherwise doesn't support ISO, should this be removed?

jfischer-no replied:
Maybe, but then it must later be added for H4 support.

Thalley replied:
Seems odd to only partially add ISO/H4 support

jfischer-no commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,507 @@
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
```
No it is not more readable.

jfischer-no commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,507 @@
+
+static void bt_hci_tx_thread(void *p1, void *p2, void *p3)
+{
+	struct usbd_class_node *const c_nd = p1;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (true) {
+		struct net_buf *bt_buf;
```
I do not see the need, the scope of c_nd (p1) is for the whole function.

Thalley replied:
Given that `p1` is in the outer scope, I will agree with you here.

jfischer-no commented:
Cherry-picked first commit from #53998 and added interface descriptor for ISO endpoints (still unused).

desowin commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,619 @@
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_in_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_if_descriptor if1_0;
```
Bluetooth specification requires 4 alternate settings with varying max packet size. Here you have only 2.

desowin replied:
It used to require in v2.1 and v3.0, but now it only recommends 4 alternate settings. Older versions can also just recommend it when they implement Core Spec Addendum 2.

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,619 @@
+
+static uint8_t bt_hci_get_int_in(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```
Minor:
```suggestion
	const struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,619 @@
+
+static uint8_t bt_hci_get_bulk_in(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```
Minor:
```suggestion
	const struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```

Thalley commented on subsys/usb/device_next/class/bt_hci.c:
```diff
@@ -0,0 +1,619 @@
+
+static uint8_t bt_hci_get_bulk_out(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```
Minor:
```suggestion
	const struct usbd_bt_hci_desc *desc = c_nd->data->desc;
```

Thalley reviewed (approved):
Some minor suggestions left, otherwise LGTM
